### PR TITLE
Add get bounding box to ElementHandle for easier element screenshot

### PR DESCRIPTION
**Pull request by elisherer**
~~This will add the ability to capture a screenshot based on a selector (as opposed to using the clip object).~~
~~The implementation is on top of the clip option to make things simple:~~

~~await page.screenshot({~~
  ~~clip: '.box:nth-of-type(43)'~~
~~});~~

~~*The example above is from the unit test (see output image below in the affected files).~~

EDIT: Changed to adding a method: `elementHandle.boundingBox()` which will return the rectangle needed for screenshot's clip option.

### CODE REVIEW

**Comment by ebidel** [comment]
FWIW, I posted [a pr](https://github.com/GoogleChrome/puppeteer/pull/452) that shows how to do this in user-code. It's just a few lines but gets even smaller after https://github.com/GoogleChrome/puppeteer/issues/382 lands.

That said, this might be common enough that it's something we should support directly in the API.

**Comment by aslushnikov** [comment]
That's nice. Let's make `elementHandle.screenshot(options)` method instead to keep `Page` lean.

**Comment by elisherer** [comment]
The challenge is that all of the screenshot code is written on the `Page` class (together with a `screenshotTaskQueue`). 
Also, `ElementHandle` does not have a reference to the creating page (maybe it should?)
Is `elementHandle.screenshot(page, options)` acceptable?

**Comment by elisherer** [comment]
I have tried to resolve the cla check by adding an alternative email of mine...

**Comment by aslushnikov** [comment]
@elisherer ah, I see. Let's do the follwing:
- we'll introduce the `elementHandle.ownerFrame()` shortly (there's [some work upstream](https://chromium-review.googlesource.com/c/chromium/src/+/624954) to make it possible)
- then you'll be able to add a `frame.page()` and use `page.screenshot` to drive element screenshots.

**Comment by elisherer** [comment]
@aslushnikov, how about we put the screenshot taking logic into one common place.
I noticed the page is only important for using the fullPage option (usint its viewport), which elements' screenshots don't need.
Also, clipping should not be possible for elements. 
So each of the cases have different "predefined" clip providing options.
In addition, the screenshot task queue is stored on the browser and passed to each of the pages. In the same way it could be passed to each element as a handle for screenshot capturing.
WDYT?

**Comment by aslushnikov** [comment]
@elisherer this approach would work. However, I don't like it because it touches quite a bit of code with no good excuse.

I see the `elementHandle.screenshot` as a convenience method for `page.screenshot`: the feature doesn't add any new capabilities and It's easy to polyfill with current puppeteer API:
```js
// Polyfill for elementHandle.screenshot
async function screenshotElement(page, selector, options)  {
  const handle = await page.$(selector);
  const box = await getElementCoordinates(handle); // roughly return element's bounding box
  options.clip = box;
  return await page.screenshot(options);
}
```
Convenience methods should be implemented atop of core, with as little footprint as possible. This will minimize the maintenance cost.

> In addition, the screenshot task queue is stored on the browser and passed to each of the pages. In the same way it could be passed to each element as a handle for screenshot capturing.

The `screenshotTaskQueue` is a shabby detail of screenshot machinery we have today, it's unfortunate to have it spreading further into subsystems.

File: test/test.js
Comment (JoelEinbinder) [inline test/test.js]: Nit: fortySecondBoxElement

File: lib/ElementHandle.js
Comment (JoelEinbinder) [inline lib/ElementHandle.js]: The bounding box of the element could be partially off-screen. The element might also not be scrolled into view. something to consider when taking screenshots. If you look above at `_visibleCenter`, we use `scrollIntoViewIfNeeded` and clip based on the window bounds. Because we are already in JavaScript, we use getBoundingClientRect and avoid introducing `DOM.getBoxModel`.

I don't expect `element.boundingBox()` to scroll the element into view, but I can imagine that it would be intended if boundingBox is used to take screenshots.

File: test/test.js
Comment (elisherer) [reply test/test.js]: Needs explaining here... You don't like the long variable name?

File: lib/ElementHandle.js
Comment (elisherer) [reply lib/ElementHandle.js]: Would you recommend documenting this consideration?

Comment (aslushnikov) [inline lib/ElementHandle.js]: Values in `DOM.getBoxModel` are affected by the `box-sizing` css property. AFAIR they also include page scale (@wwwillchen might remember better).

`getBoundingClientRect` might work better for this use case.

> Would you recommend documenting this consideration?

Once this lands, https://github.com/GoogleChrome/puppeteer/pull/452 should work as a good-enough documentation.

File: test/test.js
Comment (aslushnikov) [inline test/test.js]: let's do `it` instead of `fit`.

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: let's return nullable `box` without throwing

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: Nice catch!

Could you please change this into a more elaborate message while we're here:
```js
  throw new Error('Element is detached from DOM');
```

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: Is there any reason why you don't use the `rect.width` and `rect.height` here?

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: This returns bounding box in viewport, whereas `screenshot` accepts coordinates relative to the main frame. To make this usable with `page.screenshot`, you need to convert coordinates to the window.

That's how we do this in DevTools front-end: [DOMExtension.js](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/devtools/front_end/dom_extension/DOMExtension.js?type=cs&q=AnchorBox+f:front_end&sq=package:chromium&l=494)

Let's add a test that verifies screenshots with clipping, something along the lines:
```js
describe('Page.screenshot', function() {
  ...
  it('should work with elementHandle.boundingBox', SX(async function() {
    await page.setViewport({width: 500, height: 500});
    await page.goto(PREFIX + '/grid.html');
    await page.evaluate(() => window.scrollBy(0, 100));
    const box = await page.$('div:nth-child(3)');
    const screenshot = await page.screenshot({clip: await box.boundingBox()});
    expect(screenshot).toBeGolden('screenshot-element-bounding-box.png');
  }));
  ...
```

File: docs/api.md
Comment (aslushnikov) [inline docs/api.md]: (relative to the main frame), or `null` if element is detached from dom.

File: lib/ElementHandle.js
Comment (elisherer) [reply lib/ElementHandle.js]: I've added the test above, but I ended up using the code from [document-offset](https://github.com/timoxley/offset) for getting the absolute offset.

Comment (aslushnikov) [reply lib/ElementHandle.js]: Thanks for adding a test!

> but I ended up using the code from document-offset for getting the absolute offset.

In case of nested iframes, this will return the bounding box inside the iframe's document, not the main frame document. (can we have a test for this too?)

Comment (elisherer) [reply lib/ElementHandle.js]: Since we are running JavaScript we are are bound to CORS rules, accessing parent frame and getting its position might not be possible (it even happens when ran from the file system).
So the DevTools approach might be better to accomplish this. I'm trying to get this to run:
```js
fit('should handle nested frames', SX(async function() {
      await page.setViewport({width: 500, height: 500});
      await page.goto(PREFIX + '/frames/nested-frames.html');
      const nestedFrame = page.frames()[0].childFrames()[1];
      const elementHandle = await nestedFrame.$('div');
      const box = await elementHandle.boundingBox();
      expect(box).toEqual({ x: 28, y: 253, width: 284, height: 18 });
    }));
```
Currently errors with:
```
  Message:
    Expected $.x = 8 to equal 28.
    Expected $.y = 8 to equal 253.
```

That being said, I'm guessing that `click` and `hover` won't work with iframes either...

Comment (elisherer) [reply lib/ElementHandle.js]: UPDATE: I tried again with DOM.getBoxModel and the content quad returned is the position of the box relative to the main frame. I will update once both tests are running (scroll on main frame and nested in iframe)

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: Let's not use `DOM.getBoxModel` - it misbehaves with page zoom. It's also an experimental protocol method (and we try to minimize our use of experimental methods so that it would be easier to stabilize them in future).

The logic here is implementable with DOM API, e.g. [like this](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/devtools/front_end/dom_extension/DOMExtension.js?q=DOMExtension&sq=package:chromium&l=494). Do you have any concerns about this approach?

Comment (elisherer) [reply lib/ElementHandle.js]: Yes, as you pointed out, we need to consider elements inside iframes and this logic you suggest applies only to the frame containing the element. Using a while loop to get the parent frame for offsets might not work because of CORS policy which makes the method inconsistant.
On the other hand using the DevTools API gives us the ability to inspect the window without browser JavaScript constraints.
So it's a matter of deciding if we want to support nested element bounding boxes or not (or we do but only for same domain frames).

Comment (aslushnikov) [reply lib/ElementHandle.js]: > Using a while loop to get the parent frame for offsets might not work because of CORS policy which makes the method inconsistant.

Yes, you're right.

We're thinking about introducing the `elementHandle.ownerFrame()` method in #433. This would allow you to iterate frame structure avoiding CORS and use bullet-proof `element.getBoundingClientRect` instead of `DOM.getBoxModel`.

If this solves your concerns, I'd like to wait for the `ownerFrame` to land first and then follow-up with this PR.

Comment (elisherer) [reply lib/ElementHandle.js]: OK, will do.

**Comment by paulirish** [comment]
@elisherer any update? looks like a few people eager for this.

**Comment by aslushnikov** [comment]
@paulirish this is stuck on my implementation of `elementHandle.ownerFrame()` method (#433).

**Comment by elisherer** [comment]
@aslushnikov, It now supports iframes by using upwards traversal using the DOM api (describeNode) and getting the position inside the window using JavaScript.
(I used my own implementation of "ownerFrame", hope it's ok)

**Comment by elisherer** [comment]
why is the cla/google "waiting for status to be reported" is waiting for a couple of days? shouldn't it be instant?

**Comment by aslushnikov** [comment]
> @aslushnikov, It now supports iframes by using upwards traversal using the DOM api (describeNode) and getting the position inside the window using JavaScript.

Thanks @elisherer. I've though about this a bit more and doing multiple hops into page to resolve bounding box seems to be fragile - page might've changed during the process. This would be terrible to debug.

I now tend to agree with you on the matter of using `DOM.getBoxModel` method. I'll look into it shortly to see how it could be fixed.

> why is the cla/google "waiting for status to be reported" is waiting for a couple of days? shouldn't it be instant?

No worries, we can ignore the bot. I believe it's not very reliable.

**Comment by aslushnikov** [comment]
@elisherer: @JoelEinbinder looked into that and it turned out that:
- `getBoxModel` is wrong with non-default page's scale
- we consider page's scale to be always default in pptr

To sum up: we can use `getBoxModel` in pptr. Please, go ahead with this patch!
I'm sorry for making you wait for so long.

Also, @JoelEinbinder has a PR to fix clicking in iframes: https://github.com/GoogleChrome/puppeteer/pull/971
It would be great if you can base your work atop of that one.

**Comment by elisherer** [comment]
@aslushnikov, OK, I'm done bringing back the code of getBoxModel while merging to the new changes in ElementHandle.
I did base my code on @JoelEinbinder PR #971, so it kind of already fixes the problem there with visible center (because I made visible center call boundingBox; no need for 2 separate identical calls).

File: lib/ElementHandle.js
Comment (aslushnikov) [inline lib/ElementHandle.js]: I wouldn't expect `element.boundingBox()` to cause side effects like scrolling page. Can we drop it from here?

Comment (elisherer) [reply lib/ElementHandle.js]: Isn't it the same side effects as keyboard and touchscreen operations (no expecatations there either)?
Maybe it should be optional?
i.e.
```
element.boundingBox({ scrollIntoView: true })
```
(it will return a different bounding box if scroll was needed)

Comment (aslushnikov) [reply lib/ElementHandle.js]: > Isn't it the same side effects as keyboard and touchscreen operations (no expecatations there either)?

This only happens for mouse/touch operations. The reason is that we can't click to the off-screen location, it is (somewhat) understandable that element should be on screen to be clicked.

Whereas The `boundingBox` method is a getter, and getters are generally perceived as operations without side-effects.

> (it will return a different bounding box if scroll was needed)

Thanks, I see why you need it. I'd make an internal method `_innerBoundingBox(scrollIntoView)`, that would be used in both `_visibleCenter` and `boundingBox()`:

```js
  async _visibleCenter() {
    const box = this._innerBoundingBox(true /* scrollIntoView */);
    // .. compute center
  }
  async boundingBox() {
    return this._innerBoundingBox(false /* scrollIntoView */);
  }
```

Comment (elisherer) [reply lib/ElementHandle.js]: I agree about the above, but how do you suggest solve the problem of "off-frame" objects needing to be captured by screenshot.
I would want the behavior of scrolling before calling `boundingBox()`.
`element.scrollIntoViewIfNeeded()` like in `focus()`?

**Comment by elisherer** [comment]
@aslushnikov , as my last commit states, I added a scrollIntoViewIfNeeded method and used is as part of the visualCenter, and in the unit test of taking a screenshot with a framed scrolled, I called it just before calling boundingBox to ensure the box is visible. looks good by me.

**Comment by aslushnikov** [comment]
> as my last commit states, I added a scrollIntoViewIfNeeded method and used is as part of the visualCenter, and in the unit test of taking a screenshot with a framed scrolled, I called it just before calling boundingBox to ensure the box is visible. looks good by me.

@elisherer `element.scrollInfoViewIfNeeded` would not be used for anything other then screenshots. In this case, we can just have the `elementHandle.screenshot`. Conveniently, the `ElementHandle` now has a reference to the page: https://github.com/GoogleChrome/puppeteer/blob/f1aa18af4e713abfb170e1f9f1d3a3d378b32987/lib/ElementHandle.js#L31

Even with `elementHandle.screenshot`, we should keep the `ElementHandle.boundingBox`: it comes handy in multiple places.

**Review by aslushnikov — APPROVED** [review]
Great work, thank you.

**Comment by zhongjiewang** [comment]
if the elementhandle in a frame,  screenshot method , does it work ?

**Pull request merged**
